document.go

```diff
@@ -0,0 +1,169 @@
+package gemini
+
+import (
+	"bytes"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// DocumentBuilder allows programmatic document creation using the builder pattern.
+// DocumentBuilder supports the use of headers and footers, which combined with the body at build time.
+type DocumentBuilder struct {
+	header string
+	body   *strings.Builder
+	footer string
+}
+
+// NewDocumentBuilder creates a DocumentBuilder.
+func NewDocumentBuilder() DocumentBuilder {
+    builder := new(strings.Builder)
+	return DocumentBuilder{"", builder, ""}
```

Comment:
This may look a bit weird but this was my solution to avoid copying the builder. If there is an easier-to-read solution I'd be happy to integrate it.